### DISCUSSION TITLE
Windows Terminal Preview v1.22.2362.0

### DISCUSSION THREAD

**Discussion by PankajBhojwani** [microsoft/terminal#17809] [Announcements]
This is, perhaps, the largest release of Windows Terminal to date. I can't believe we're here, but we finally have support for Sixels and a new console hosting infrastructure (stay tuned for a NuGet package!) and more. Read up, and read down, to find out more.

## Features

- Terminal and conhost now support Sixels! 🎤🫳 (#17421) (thanks @j4james!) (#17724) (thanks @j4james!) (#17581) (#17636)
- We have rewritten how hosted console applications are translated for Windows Terminal (and other ConPTY consumers) to be based on direct API->VT translation rather than "rendering" a point-in-time snapshot of the display. This should result in improved throughput, reliability, correctness, plus an ability for applications to send unmodified VT directly to the terminal, but may result in some compatibility issues. Please file any issues you encounter with console applications. (#17510) (#17645) (#17741) (#17757) (#17748) (#17442) (#17573)
   - ... terminals which support DEC rectangular operations `DECCRA` and `DECFRA` should see higher fidelity app-based scrolling (#17747)
- We now support a much wider range of Unicode, including Emoji with ZWJ, combining characters, flag sequences, and anything else that requires proper Grapheme Clusters! Applications can query support for grapheme clusters with `DECRPM` `2027`, and users can now configure whether Unicode is measured like the vintage Windows Console, like Linux and macOS terminals using `wcwidth`, or with Grapheme Clusters like Contour and other modern terminal emulators. Global setting `compatibility.textMeasurement` (enum `console`, `wcswidth`, `graphemes` (default)) (#16916) (#17443)
- For the first time in decades, we've changed the popup UI (used for <kbd>F7</kbd> and friends) in CMD! It now draws beneath the prompt (pushing the screen up) and has a modern TUI scrollbar if it doesn't all fit on the screen. (#17445) (#17668) (#17556)
- Terminal will now offer relevant packages from WinGet if you're running CMD on Windows 24H2 and run a nonexistent command (#16848) (#17614)
- You can now add actions like `splitPane` and `sendInput: foo` and `quit` directly to the New Tab dropdown menu! (#17281)
- You can now open a pane containing snippets (any `sendInput` actions!) with the `openSnippets` action! You can quickly access any commands you've saved in your JSON file, or with the new `x-save` command, and replay them with a new slick UI (#17330) (#17794) (#17528)
   - ... snippets which can also be loaded from a `.wt.json` file in your current working directory (we're still working on docs, please stay tuned!) (#17388)
   - ... snippets which can also be saved directly from the command line with `wt x-save "text to save"` (#16513) (thanks @e82eric!) with additional fixes (#17758)
- <kbd>Ctrl+Shift+Period</kbd> will now open the "quick actions" flyout, which contains snippets and suggestions. You can now also bind the `quickFix` action to open a menu containing only WinGet suggestions (as of today, on Windows 24H2) (#17502)
- You can now search with regular expressions in Terminal (and conhost!) (#17316)

## Changes

- `ReadConsoleOutput` and `WriteConsoleOutput` will now handle out-of-bounds conditions as documented rather than as-made-up (#17567)
- Actions now have their own IDs, and key bindings just point at them; we've made changes to how this works to improve the action and binding experience in future versions (#17162) (#17215)
- Command/input previews will now display in line with the cursor, in italic, in the right font (#17386) (#17777) (#17533)
   - ... which you can display in RGB with the fun `{profile}.experimental.rainbowSuggestions` setting (boolean, default `false`) (#17416)
- Commands displayed in the suggestions UI can now include descriptions (#17376)
- Multi-cell glyph inputs will no longer be assumed to consume two code units (#17680)
- Pixel shader authors rejoice! If your shader fails to compile, you'll now get a useful error message... (#17436) (thanks @blitzRahul!)
- Terminal will now no longer correct `SGR 37` and `SGR 40` when emitted by `powershell.exe`; this may result in regions of incorrect colors (#17666) (#17685) (additional thanks @j4james for some cleanup!)
- Vintage console applications will no longer get corrupted responses from ReadConsoleOutput when Raster Fonts are enabled (#17681)
- We have brought the new Windows 24H2 API out to the ConPTY NuGet package (#17704)
- We have enabled "automatically mark prompts" (for shell integration) by default (#17761)
- We're now collecting data on which settings and actions folks are using (#17718) (#17678)
- We're trying something new, which is putting a short version of these release notes on the Windows Store (#17730)
- We've changed the communication mechanism for the Default Terminal feature to allow the use of overlapped I/O (#17575)
- We've reintroduced support for <kbd>Alt+Numpad</kbd> input of extended off-keyboard characters (#17637) (#17774)
- When we reload a saved session, we will now display some of its contents on the screen (#17334)

### UI

- We've rewritten how selection colors work! They will now be drawn with appropriate contrast, at the exact color you specify, with a foreground that is black or white (whichever is most visible!) (#17753) (#17725) (#17638) (#17676)
- The search box will no longer cover up search results in the top right (unless there's nowhere to scroll) (#17516) (thanks @e82eric!)
- Due to a build issue, we had to disable full optimization for this build. We apologize for any potential performance regressions. (#17765)
- Terminal now comes with shiny new icons for the Visual Studio Dev Shells (#17706)
- We have restored the `...`s in some command palette names to help users find them (#17715)
- You can now configure the theme for the Settings UI separately from the application with `{theme}.settings.theme` (enum `light`, `dark`, `system`; default: `system`) (#16479) (thanks @bundgaard!)

### VT Features

- You can now configure the "Answerback" message sent in response to `ENQ` using the profile setting `answerbackMessage` (string, default: `null`) (conhost users: `HKCU\Console` `AnswerbackMessage` `REG_SZ`) (#17660) (thanks @j4james!)
- We now support the DEC VT paging operations, `NP` `PP` `PPA` `PPR` and `PPB`, as well as `DECPCCM` (page cursor coupling mode) and `DECRQDE` (request displayed extent) (#16615) (#17353) (thanks @j4james!)
- Applications can now request the entire color palette using `DECRQTSR` `DECCTR` (`CSI 1 ; Pt $ u`) (#17708) (thanks @j4james!)
- Applications can now query the cursor shape set with `DECSCUSR` using `DECRQSS` (`DCS $ q SP q ST`) (#17659) (thanks @j4james!)
- Applications can now query the size of character cells (`CSI 16 t`) and the entire text area (`CSI 14 t`), with a response measured in pixels (#17504) (thanks @j4james!)
- Applications can now query the color palette with `OSC 4` `?`, and the default foreground (`OSC 10 ?`), background (`OSC 11 ?`), cursor (`OSC 12 ?`) and selection colors (`OSC 17 ?`) (#17729)
- Applications can now change the selection color using `OSC 17` (#17742)
- Reports generated in response to VT requests are once again given priority over other input (#17786)

## Bug Fixes

- We will no longer conveniently forget all the colors on the last line of the screen when you resize it (#17568)
- Terminal will now more reliably save your session when your computer reboots to update (#17714)
- Background colors will no longer bleed over the edges of the screen when D2D is in use (such as over RDP to a VM) (#17674)
- conhost: pasting a file from File Explorer will no longer result in two copies of the path (#17543)
- conhost: when we generate <kbd>Alt+Numpad</kbd> events to input clipboard contents on Windows, we will now no longer mistranslate CP936 text (#17635)
- Curly underlines will now render properly on double-width `DECDWL` and double-height `DECDHL` lines (#17679)
- cursor invalidation once again works properly on double-width or double-height lines (#17234) (thanks @j4james!)
- cursor movement over ConPTY now works properly on double-width or double-height lines (#17233) (thanks @j4james!)
- The rendering of curly underlines has been improved (#17501) (thanks @smprather for reporting!) (#17508)
- The Suggestions UI will no longer try to grab the PowerShell "ghost text" out of your input line (#17781)
- You can no longer insert a `saveSnippet` action in your JSON or bind it to any keys (#17535)
- You can now successfully input more than one Emoji at once on Windows 10 (#17213)
- You can once again cancel overwriting a key binding in the Settings UI (#17763)
- You can once again disable the detection of URLs (#17731)
- You can once again use <kbd>Ctrl+D</kbd> and <kbd>Enter</kbd> in a disconnected pane (#17750)

### Accessibility

- Accessibility API `RangeFromPoint` and `ExpandToEnclosingUnit` will no longer cause Terminal to hang (#17695)

### Performance

- Scrolling the screen will now result in less GPU traffic (due to erroneous invalidations) (#17583)
- Terminal now makes fewer copies (and heap allocations) when writing input to the console app; it's also more correct! (#17710)
- The command palette has lost some extraneous animations, so it should feel much faster (#17335)

### Reliability

- AtlasEngine will no longer over-read a memory buffer when setting up a GPU texture (#17536)
- Closing a window when you have an `unfocusedAppearance` set will no longer result in unscheduled disassembly of your terminal (#17770)
- Overall reliability *should* be improved by us having better exception handling (#17783)
- Session restoration now works more reliably when the entire scrollback is full (#17654)
- We've fixed an instance of Terminal crashing during settings reload (#17751)
- You can no longer trivially crash Terminal by holding down <kbd>Enter</kbd> while spawning new tabs (#17788)

### VT Compatibility

- ConPTY: responses longer than 4kb will no longer be corrupted (#17738) (thanks @j4james!)
- VT mouse events will once again be generated when your mouse is off the top of the terminal window (and as a bonus, we'll no longer autoscroll!) (#17779)

### Code Health

- Resize with reflow is, like, way faster in Debug builds (#17550)
- We added `reloadEnvironmentVariables` to our oft-forgotten JSON schema (#17696)
- We've made some changes to the shell extension and elevation shim to make them easier to maintain (#15327)
- build: all of our open-source dependencies have been migrated to vcpkg (#15855)
- build: you can now launch Windows Terminal with <kbd>F5</kbd> from Visual Studio (#17577)
- conpty: if your terminal fails to respond to `DSR` `CPR`, we'll continue without it in 3 seconds (#17576) (#17574)
- We've done some tidying on pane content to ensure that our implementations don't diverge (#17537)
- We've made some improvements to how we handle movement across rectangles for copy operations (#17143)

---

_Binary files inside the unpackaged distribution archive bear the version number `1.22.240823002-preview`._

### Asset Hashes

- Microsoft.WindowsTerminalPreview_1.22.2362.0_8wekyb3d8bbwe.msixbundle
   - SHA256 `E00B06D4197E6002E9B92A076044F12B105956ABAF9DE058D9B3DF021B8A18B4`
- Microsoft.WindowsTerminalPreview_1.22.2362.0_arm64.zip
   - SHA256 `8F3A10ED10980379277127037FB4A66F40E1A113EDF83991B6FF7B2D24613BA5`
- Microsoft.WindowsTerminalPreview_1.22.2362.0_x64.zip
   - SHA256 `6EA8DAF670A28DD43823079617E2054A9A5F4411D0B7025C637B6C6276621D81`
- Microsoft.WindowsTerminalPreview_1.22.2362.0_x86.zip
   - SHA256 `AB59CFB83E645012733B13BF9C155C7F29D70E3DE75886D46F1CFAF64E90E915`

<hr /><em>This discussion was created from the release <a href='https://github.com/microsoft/terminal/releases/tag/v1.22.2362.0'>Windows Terminal Preview v1.22.2362.0</a>.</em>

**Comment by ofek** [comment]
How can one try out Sixel support? I can't quite find a program to play videos or display pictures in the terminal.

**Reply by zadjii-msft** [reply]
[`lsix`](https://github.com/hackerb9/lsix) is fantastic:

![image](https://github.com/user-attachments/assets/245fed63-5a65-4ca0-9558-113b23517330)

**Reply by ofek** [reply]
Thanks! Is there a Windows-native tool I can use?

**Reply by Andarwinux** [reply]
mpv supports sixel, but since [shinchiro's build](https://github.com/shinchiro/mpv-winbuild-cmake) doesn't have sixel, you'll need to compile it yourself, or my [personal build](https://github.com/Andarwinux/mpv-winbuild).

**Reply by ofek** [reply]
I just tried the following command but only audio plays no video and I cannot quit the audio without using Task Manager: `mpv --really-quiet --vo=sixel video_1280.mp4`

**Reply by Andarwinux** [reply]
You need put mpv.com next to mpv.exe

**Reply by ofek** [reply]
It is shipped that way, no?

![image](https://github.com/user-attachments/assets/9b60aacb-c200-431b-91c8-ea58c25ed80a)

**Reply by Andarwinux** [reply]
That's very strange, to me it only happens when mpv.com doesn't exist. Does this work with other vo like gpu-next?

**Reply by ofek** [reply]
Yes that works and opens up the actual window.

**Reply by Andarwinux** [reply]
It looks like my mpv.com doesn't work on your machine, maybe you can try replacing it with shinchiro's? Also, are you sure you're executing mpv.com and not mpv.exe?

**Reply by ofek** [reply]
My apologies, I didn't know I had to use that executable. Why is that? It works! Although I get a cascade of vertical videos (which I'm assuming represent each frame) rather than what I would expect as a single video object. Is that expected?

**Reply by Andarwinux** [reply]
> My apologies, I didn't know I had to use that executable. Why is that? It works!

This is a Windows limitation, see https://devblogs.microsoft.com/oldnewthing/20090101-00/?p=19643

> Although I get a cascade of vertical videos (which I'm assuming represent each frame) rather than what I would expect as a single video object. Is that expected?

This is a known issue, you need to manually specify a large enough `vo-sixel-width`

**Reply by ofek** [reply]
Fantastic, thank you very much!

**Reply by DHowett** [reply]
> This is a Windows limitation

I have good news! We finally fixed that.

**Reply by Andarwinux** [reply]
> > This is a Windows limitation
> 
> I have good news! We finally fixed that.

What's that?

**Reply by zadjii-msft** [reply]
> > I have good news! We finally fixed that.
> 
> What's that?

He's referencing #7337, which I _think_ shipped in Windows [26052](https://blogs.windows.com/windows-insider/2024/02/08/announcing-windows-11-insider-preview-build-26052-canary-and-dev-channels/), though I'm pretty sure we're backporting it to win10 too. Not sure we've got a blog post about that specifically though - we should!

**Reply by Andarwinux** [reply]
> He's referencing https://github.com/microsoft/terminal/pull/7337, which I think shipped in Windows [26052](https://blogs.windows.com/windows-insider/2024/02/08/announcing-windows-11-insider-preview-build-26052-canary-and-dev-channels/),

Good to know, I tried this on Canary and it worked great!

> though I'm pretty sure we're backporting it to win10 too. Not sure we've got a blog post about that specifically though - we should!

Does backport apply to all Windows 10 versions? When will it be rolled out? For mpv, we still need support for Windows 10 1607.

**Reply by zadjii-msft** [reply]
> Does backport apply to all Windows 10 versions

Definitely not. I believe (I am probably mistaken) that 19041 is the build that gets regular servicing, but the bar for bringing things back farther than that is much, much higher (and this probably wouldn't meet that criteria)

**Reply by lhecker** [reply]
FYI you can find the preliminary documentation here: https://github.com/MicrosoftDocs/Console-Docs/pull/323

**Comment by smprather** [comment]
Am I just looking too soon?

```
PS C:\Users\******> winget list
Name                                                           Id                                                     Version                  Available        Source
----------------------------------------------------------------------------------------------------------------------------------------------------------------------
Windows Terminal Preview                                       Microsoft.WindowsTerminal.Preview                      1.21.1772.0                               winget
```

**Reply by smprather** [reply]
This is resolved, BTW. New version shows up now.

**Comment by sxyazi** [comment]
It seems like [DA1](https://vt100.net/docs/vt510-rm/DA1.html) is broken in this release. When I query Sixel support by sending `\x1b[c`, the response doesn't include a valid CSI (`\x1b`) in the beginning:

```
[?61;4;6;7;14;21;22;23;24;28;32;42c
```

and it should be:

```
\x1b[?61;4;6;7;14;21;22;23;24;28;32;42c
```

I'm using [Microsoft.WindowsTerminalPreview_1.22.2362.0_arm64.zip](https://github.com/microsoft/terminal/releases/download/v1.22.2362.0/Microsoft.WindowsTerminalPreview_1.22.2362.0_arm64.zip)

**Reply by zadjii-msft** [reply]
I promoted this out to #17813 to make sure it gets its own investigation, thanks!

**Comment by AbraxasDMG** [comment]
Hi everyone.

I saw the following statement in the Preview v1.22 release notes:

> ⚡ Actions like splitPane, sendInput: someinput, and quit can now be added directly to the New Tab dropdown menu

I am trying to find __any__ information on how to do that. Could anyone point me in the right direction?
Thanks in advance!

**Reply by zadjii-msft** [reply]
Op, good catch. Looks like we forgot to update https://learn.microsoft.com/en-us/windows/terminal/customize-settings/appearance#new-tab-dropdown

**Reply by zadjii-msft** [reply]
Filed https://github.com/MicrosoftDocs/terminal/issues/794 to track this. 

----

Say you have an action like 

```json

        {
            "command":
            {
                "action": "sendInput",
                "input": "git log -10 --pretty=oneline --abbrev-commit"
            },
            "name": "git log -10",
            "id": "gitlog10"
        },
```

then you can add it to the menu with:

```json
    "newTabMenu": 
    [
        {
            "type": "remainingProfiles"
        },
        {
            "id": "gitlog10",
            "type": "action"
        }
```

The `id` is the bit that's relevant. Just make sure the `id` in the `newTabMenu` matches the one in the `actions`

**Comment by VPraharsha03** [comment]
Big shout out to the devs! I was eagerly waiting to try this feature out.

Just tried out sixel support and compared them on the Windows terminal and OpenConsole, the feature seems to be complete on Windows Terminal but on the OpenConsole images produced are scaled higher and have jagged edges no matter what the resolution is,

command ran was `img2sixel ~/kirby.png -w 350px -h 300px`

Windows Terminal (1.22.240823002-preview)

![kirby](https://github.com/user-attachments/assets/10878619-650c-4bf0-b0fb-836b8bae0044)

Open Console:

![image](https://github.com/user-attachments/assets/68db83fd-65d8-48f4-9ebc-0daf282da386)

**Reply by j4james** [reply]
The image is scaled to match the text, so the difference in size is assumedly because you're using different fonts or different font sizes. It should occupy the same number of text rows and columns in both terminals.

I'm not sure about the jagged edges, but that might just be a difference between the GDI renderer used in OpenConsole and the DirectX renderer use in Windows Terminal. This may not be officially supported, but you can enable the DirectX renderer in OpenConsole by going to the registry and creating a `UseDx` DWORD value inside the `HKEY_CURRENT_USER\Console` key, and setting the value to 1.

**Reply by VPraharsha03** [reply]
Thanks for the tip! Using DirectX renderer smoothened the edges

**Comment by damnskippy** [comment]
Just wondering - is there a reason (like minimum OS version or something) why Preview version installed via Microsoft Store wouldn't offer an upgrade or automatically upgrade to the latest version?  I'm currently running preview version 1.21.1772.0, and while Store page shows the version 1.22.2362.0 as available including the section about "what's new in this version", it just doesn't offer an update option?   FWIW, winver on my machine shows Version 22H2 (0S Build 22621.4037).  Can I just force the upgrade by downloading the latest artifacts from the release page, would that cause a conflict with the Store installed version?  Thanks.

**Reply by lhecker** [reply]
The version you download from the store is the exact same as from our release page and they will not conflict with each other. If you ever install Windows Terminal from our release page on GitHub, it'll even get updates from the Microsoft Store app as usual.

As to the reason why 1.22 isn't available yet: I believe we haven't rolled out the update in the store yet, but the store translations have already been updated, which includes the changelog.

**Reply by damnskippy** [reply]
Thank you!

**Reply by DHowett** [reply]
UGH. Nope, that was just me failing to push the packages at the same time as the release notes. I'm sorry.

**Reply by damnskippy** [reply]
No worries, and thanks @DHowett

**Comment by alvin-qh** [comment]
"Clear Buffer" command not work on this version? I execute "Clear Buffer" command but nothing was happened

**Reply by j4james** [reply]
It sounds like you're encountering the same problem as issue #17867.

**Comment by 0wwafa** [comment]
I tested OpenConsole.exe and I see quite some problems:
1) restore defaults has no effects.
2) the scroll back is very short (a page and a half) and there is no way to set it.
3) the fonts look smaller than in the original conhost

**Comment by VPraharsha03** [comment]
when can we expect this to come into final release ?
